typescript/sdk/src/providers/SmartProvider/SmartProvider.test.ts

```diff
@@ -0,0 +1,385 @@
+      const provider = new TestableSmartProvider([provider1, provider2]);
+
+      try {
+        await provider.simplePerform('getBlockNumber', 1);
+        expect.fail('Should have thrown an error');
+      } catch (e: any) {
+        expect(e.name).to.equal('Error');
+        expect(e.isRecoverable).to.be.undefined;
+        expect(e.cause).to.equal(serverError1); // First error should be the cause
+        expect(provider1.called).to.be.true;
+        expect(provider2.called).to.be.true;
+      }
+    });
+
+    it('both providers timeout, combined timeout error is thrown', async () => {
+      const provider1 = MockProvider.success('success1', 2000);
+      const provider2 = MockProvider.success('success2', 2000);
+      const provider = new TestableSmartProvider([provider1, provider2]);
+
+      try {
+        await provider.simplePerform('getBlockNumber', 1);
+        expect.fail('Should have thrown an error');
+      } catch (e: any) {
+        expect(e.name).to.equal('Error');
+        expect(e.isRecoverable).to.be.undefined;
+        expect(e.message).to.include('All providers timed out');
+        expect(provider1.called).to.be.true;
+        expect(provider2.called).to.be.true;
+      }
+    });
+
+    it('blockchain error stops trying additional providers immediately', async () => {
+      const blockchainError = new ProviderError(
+        'execution reverted',
+        EthersError.CALL_EXCEPTION,
+      );
+
+      const provider1 = MockProvider.error(blockchainError);
+      const provider2 = MockProvider.success('success2');
+      const provider = new TestableSmartProvider([provider1, provider2]);
+      try {
+        await provider.simplePerform('getBlockNumber', 1);
+        expect.fail('Should have thrown an error');
+      } catch (e: any) {
+        expect(e.name).to.equal('BlockchainError');
+        expect(e.isRecoverable).to.equal(false);
+        expect(e.message).to.equal('execution reverted');
+        expect(e.cause).to.equal(blockchainError);
+        expect(provider1.called).to.be.true;
+        expect(provider2.called).to.be.false; // Key test - second provider should NOT be called
+      }
+    });
+
+    it('blockchain error takes priority over server error in actual flow', async () => {
+      const serverError = new ProviderError(
+        'connection refused',
+        EthersError.SERVER_ERROR,
+      );
+      const blockchainError = new ProviderError(
+        'insufficient funds',
+        EthersError.INSUFFICIENT_FUNDS,
+      );
+
+      const provider1 = MockProvider.error(serverError);
+      const provider2 = MockProvider.error(blockchainError);
+      const provider = new TestableSmartProvider([provider1, provider2]);
+
+      try {
+        await provider.simplePerform('getBlockNumber', 1);
+        expect.fail('Should have thrown an error');
+      } catch (e: any) {
+        expect(e.name).to.equal('BlockchainError'); // Should get blockchain error, not server error
```

Comment:
why not check by validating that the error is an instance of the `BlockchainError` instead?